libyul/optimiser/Suite.cpp

```diff
@@ -429,6 +244,50 @@ map<char, string> const& OptimiserSuite::stepAbbreviationToNameMap()
 	return lookupTable;
 }
 
+void OptimiserSuite::runSequence(string const& _stepAbbreviations, Block& _ast)
+{
+	bool insideLoop = false;
+	vector<string> steps;
+	for (char abbreviation: _stepAbbreviations)
+	{
+		auto const& step = stepAbbreviationToNameMap().find(abbreviation);
+		if (step != stepAbbreviationToNameMap().end())
+			steps.push_back(step->second);
+		else if (abbreviation == ' ' || abbreviation == '\n')
+		{
+			// Ignore spaces and newlines added for readability
+		}
+		else if (abbreviation == '(')
+		{
+			assertThrow(!insideLoop, OptimizerException, "Nested parentheses not supported");
+
+			insideLoop = true;
+			if (steps.size() > 0)
+			{
+				runSequence(steps, _ast);
```

Comment:
It looks a bit weird to me that actually running the steps only happens when you encounter the first `(`.

Maybe it's better to preprocess the string first (search for '(' etc)?

Reply:
Well, I could just have a helper that finds the next `(`, `)` or end of string and returns a substring for me to execute. Then run it in a loop. This would let me get rid of some `if`s, but not all of them. And I would have to separately remove spaces.

A different solution would be to preprocess it all first and store sequence fragments in a vector. I would also need one for flags to indicate which fragments were in parentheses. I don't think it's simpler.

The current approach is just a simple state machine. Processes the string in one pass, without backtracking. I could make it look more like a real FSM by changing names a bit, having proper states instead of extra boolean flags and extracting the part that consumes the output into a separate object. I didn't want to overengineer it though.

I think I prefer the current approach but I can change it if you think a different one would be better.

Reply:
I don't think that storing each step in a vector is needed, but I think it is good to extract the loop- and non-loop-part. You can use `erase_all(steps, " ")` to remove the spaces.

The problem with the state machine is that individual parts are hard to understand unless you see the full machine.

Reply:
I have rewritten the function. It now has separate validation and processing loops and strips whitespace upfront to make those parts simpler.

It added some duplication so I put the repeating part of code in a lambda. Is that ok or should I put it outside? It could just as well be a member function.